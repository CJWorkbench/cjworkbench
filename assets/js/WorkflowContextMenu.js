// Drop-down menu on Workflows page, 
// triggered by click on three-dot icon next to listed workflow

import React from 'react';
import { csrfToken, goToUrl } from './utils';
import { UncontrolledDropdown, DropdownToggle, DropdownMenu, DropdownItem } from 'reactstrap';

export default class WorkflowContextMenu extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      workflows: []
    };
  }
  
  deleteOption(id) {
    console.log("Clicked the Delete selection in WorkflowContext Menu for ID: " + id);
    this.deleteWorkflow(id);
  }

  // Ask the user if they really wanna do this. If sure, post DELETE to server
  // TO FIX: Does not automatically reload page with new WF list after deletion 
  deleteWorkflow(id) {
    if (!confirm("Permanently delete this workflow?"))
      return;
    var _this = this;

    fetch(
      '/api/workflows/' + id ,
      {
        method: 'delete',
        credentials: 'include',
        headers: {
          'X-CSRFToken': csrfToken
        }
      }
    )
    .then(response => {
      if (response.ok) {
        var workflowsMinusID = this.state.workflows.filter(wf => wf.id != id);
        _this.setState({workflows: workflowsMinusID, newWorkflowName: this.state.newWorkflowName})
      }
    })
  }

  // \u22EE = three-dot menu icon in Unicode 
  render() {
    return (
       <UncontrolledDropdown>
        <DropdownToggle>
          {'\u22EE'}
        </DropdownToggle>
        <DropdownMenu right >
          <DropdownItem key={1} onClick={() => this.deleteOption(this.props.id)}>                       
            Delete 
          </DropdownItem>
        </DropdownMenu>
       </UncontrolledDropdown>
    );
  }
}

// type checking
WorkflowContextMenu.propTypes = {
  id: React.PropTypes.number,  
};

