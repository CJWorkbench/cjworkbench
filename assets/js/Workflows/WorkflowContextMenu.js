// Drop-down menu on Workflows List page, for each listed WF
// triggered by click on three-dot icon next to listed workflow

import React from 'react'
import PropTypes from 'prop-types'
import { UncontrolledDropdown, DropdownToggle, DropdownMenu, DropdownItem } from '../components/Dropdown'
import { Trans } from '@lingui/macro'

export default class WorkflowContextMenu extends React.Component {
  static propTypes = {
    workflowId: PropTypes.number.isRequired,
    deleteWorkflow: PropTypes.func, // func(id) ... or null if cannot delete
    duplicateWorkflow: PropTypes.func.isRequired
  }

  handleClickDelete = () => {
    const { workflowId, deleteWorkflow } = this.props
    if (deleteWorkflow) deleteWorkflow(workflowId)
  }

  handleClickDuplicate = () => {
    const { workflowId, duplicateWorkflow } = this.props
    duplicateWorkflow(workflowId)
  }

  render () {
    return (
      <UncontrolledDropdown>
        <DropdownToggle className='context-button'>
          <i className='icon-more' />
        </DropdownToggle>
        <DropdownMenu>
          <DropdownItem onClick={this.handleClickDuplicate} className='duplicate-workflow'>
            <i className='icon-duplicate' />
            <span><Trans id='js.Workflows.WorkflowContextMenu.duplicate'>Duplicate</Trans></span>
          </DropdownItem>
          {this.props.deleteWorkflow ? (
            <DropdownItem onClick={this.handleClickDelete} className='delete-workflow'>
              <i className='icon-bin' />
              <span><Trans id='js.Workflows.WorkflowContextMenu.delete'>Delete</Trans></span>
            </DropdownItem>
          ) : null}
        </DropdownMenu>
      </UncontrolledDropdown>
    )
  }
}
