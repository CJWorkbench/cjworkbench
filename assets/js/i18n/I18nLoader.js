import React from 'react'
import { I18nProvider } from '@lingui/react'
import fetchCatalog from './catalogs'

export class I18nLoader extends React.Component {
  state = {
    catalogs: {},
  }
  
  loadCatalog = (language) => {
      this.setState(state => ({
          catalogs: {
            ...state.catalogs,
            [language]: fetchCatalog(language)
          }
      }))
  }
  
  componentDidMount() {
    this.loadCatalog(this.props.language)
  }

  shouldComponentUpdate(nextProps, nextState) {
    const { language } = nextProps
    const { catalogs } = nextState

    if (language !== this.props.language && !catalogs[language]) {
      this.loadCatalog(language)
      return false
    }

    return true
  }

  render () {
    const { children, language } = this.props
    const { catalogs } = this.state

    // Skip rendering when catalog isn't loaded.
    if (!catalogs[language]) {
        return null
    }

    return (
      <I18nProvider language={language} catalogs={catalogs}>
        {children}
      </I18nProvider>
    )
  }
}
