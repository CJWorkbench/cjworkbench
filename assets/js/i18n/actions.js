const SET_LOCALE = 'SET_LOCALE';

function reduceSetLocale(state, action){
  return {
    ...state,
    locale: action.locale
  };
}

export function setLocaleAction(locale){
  return {
    type: SET_LOCALE,
    locale: locale
  }
}

export const reducerFunctions = {
  [SET_LOCALE]: reduceSetLocale
}

export function localeReducer(state, action){
  if (action.type in reducerFunctions) {
    // Run a registered reducer
    return reducerFunctions[action.type](state, action)
  }
  
  return state;
}
