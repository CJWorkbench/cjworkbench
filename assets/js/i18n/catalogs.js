import enCatalog from '../../locale/en/messages.po'
import elCatalog from '../../locale/el/messages.po'

const catalogs = {
  en: enCatalog,
  el: elCatalog
}

export default function fetchCatalog (language) {
  return catalogs[language]
}
