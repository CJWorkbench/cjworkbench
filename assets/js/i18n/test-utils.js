import { I18nProvider } from '@lingui/react'
import React from 'react'
import { shape, object } from 'prop-types'
import { mount, shallow } from 'enzyme'

/*
 * Below lie utils for testing with i18n.
 * See https://lingui.js.org/guides/testing.html
 */
// Create the I18nProvider to retrieve context for wrapping around.
const intlProvider = new I18nProvider({
  language: 'en',
  catalogs: {
    en: {}
  }
}, {})

const {
  linguiPublisher: {
    i18n: originalI18n
  }
} = intlProvider.getChildContext()

// You customize the i18n object here:
export const i18n = {
  ...originalI18n,
  // provide _ macro, for trying to unwrap default/id or just passing down the key
  _: key => {
    if (typeof key.defaults === 'string') return key.defaults
    if (typeof key.id === 'string') return key.id
    return key
  }
}

/**
 * When using Lingui `withI18n` on components, props.i18n is required.
 */
function nodeWithI18nProp (node) {
  return React.cloneElement(node, { i18n })
}

/*
 * Methods to use
 */
export function shallowWithI18n (node, { context } = {}) {
  return shallow(
    nodeWithI18nProp(node),
    {
      context: Object.assign({}, context, { i18n })
    }
  )
}

export function mountWithI18n (node, { context, childContextTypes } = {}) {
  const newContext = Object.assign({}, context, { linguiPublisher: { i18n } })
  /*
   * I18nProvider sets the linguiPublisher in the context for withI18n to get
   * the i18n object from.
   */
  const newChildContextTypes = Object.assign({},
    {
      linguiPublisher: shape({
        i18n: object.isRequired
      }).isRequired
    },
    childContextTypes
  )
  return mount(
    nodeWithI18nProp(node),
    {
      context: newContext,
      childContextTypes: newChildContextTypes
    }
  )
}
