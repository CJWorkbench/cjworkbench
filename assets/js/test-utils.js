/* globals jest */
import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import promiseMiddleware from 'redux-promise-middleware'
import errorMiddleware from './error-middleware'
import { workflowReducer } from './workflow-reducer'

// Returns new mock function that returns given json. Used for mocking "get" API calls
export function jsonResponseMock (json) {
  return jest.fn().mockImplementation(() =>
    Promise.resolve(json)
  )
}

// Returns new mock function that gives an OK HTTP response. Use for mocking "set" API calls
export function okResponseMock () {
  return jsonResponseMock(null)
}

// Helper function to return a promise that resolves after all other promise mocks,
// even if they are chained like Promise.resolve().then(...).then(...)
// Technically: this is designed to resolve on the next macrotask
// https://stackoverflow.com/questions/25915634/difference-between-microtask-and-macrotask-within-an-event-loop-context
export function tick () {
  return new Promise(resolve => {
    setTimeout(resolve, 0)
  })
}

export function sleep (ms) {
  return new Promise(resolve => {
    setTimeout(resolve, ms)
  })
}

/**
 * Return `[ notify, promise ]`; `promise` completes when `notify` is called.
 *
 * Usage:
 *
 *     const [ notify, condition ] = createConditionVariable
 *     const done = condition.then(do_something)
 *     await Promise.all([sleep(0.1).then(notify), done])
 */
export function createConditionVariable () {
  let notify
  const promise = new Promise(resolve => { notify = resolve })
  return [notify, promise]
}

/**
 * Mock Redux store with in-production reducers.
 *
 * This is for integration-style tests: they test the reducer _and_ the caller
 * at the same time.
 *
 * Usage:
 *
 *     const api = { undo: jest.fn() }
 *     const store = mockStoreWithReducer({'workflow': ..., ...}, api)
 *     store.dispatch(Actions.undo())
 *     expect(api.undo).toHaveBeenCalled()
 *     expect(store.getState().workflow).toEqual(...)
 */
export function mockStore (initialState, api = {}) {
  const middlewares = [errorMiddleware(), promiseMiddleware, thunk.withExtraArgument(api)]
  const store = createStore(workflowReducer, initialState, applyMiddleware(...middlewares))
  return store
}
