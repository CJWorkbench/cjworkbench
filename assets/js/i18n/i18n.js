import { setupI18n } from "@lingui/core"
import { currentLocale } from './locales'
import { fetchCatalog } from './catalogs'

export default const i18n = setupI18n({
   language: currentLocale,
   catalogs: {
       currentLocale: fetchCatalog(currentLocale)
   }
})
