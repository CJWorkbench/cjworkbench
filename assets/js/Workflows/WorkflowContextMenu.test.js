/* globals afterEach, beforeEach, describe, expect, it, jest */
import React from 'react'
import { act } from 'react-dom/test-utils'
import { fireEvent } from '@testing-library/react'
import { renderWithI18n } from '../i18n/test-utils'
import WorkflowContextMenu from './WorkflowContextMenu'

describe('WorkflowContextMenu', () => {
  let globalConfirm
  beforeEach(() => {
    globalConfirm = global.confirm
    global.confirm = jest.fn()
  })
  afterEach(() => {
    global.confirm = globalConfirm
  })

  it('deletes workflow', async () => {
    const workflow = { id: 123 }
    let resolvePromise
    const promise = new Promise(resolve => { resolvePromise = resolve })
    const api = {
      deleteWorkflow: jest.fn(() => promise),
      duplicateWorkflow: jest.fn(),
      updateAclEntry: jest.fn(),
      deleteAclEntry: jest.fn(),
      setWorkflowPublic: jest.fn()
    }
    const onWorkflowChanging = jest.fn()
    const onWorkflowChanged = jest.fn()

    const { getByText, getByTitle } = renderWithI18n(
      <WorkflowContextMenu
        workflow={workflow}
        api={api}
        onWorkflowDuplicating={jest.fn()}
        onWorkflowDuplicated={jest.fn()}
        onWorkflowChanging={onWorkflowChanging}
        onWorkflowChanged={onWorkflowChanged}
      />
    )

    global.confirm.mockReturnValue(true)

    fireEvent.click(getByTitle('menu'))
    fireEvent.click(getByText('Delete'))

    expect(onWorkflowChanging).toHaveBeenCalledWith(123, { isDeleted: true })
    expect(api.deleteWorkflow).toHaveBeenCalledWith(123)
    expect(onWorkflowChanged).not.toHaveBeenCalled()
    resolvePromise(null)
    await act(async () => await null) // respond to resolvePromise(null)
    expect(onWorkflowChanged).toHaveBeenCalledWith(123)
  })

  it('duplicates workflow', async () => {
    const workflow = { id: 123 }
    let resolvePromise
    const promise = new Promise(resolve => { resolvePromise = resolve })
    const api = {
      deleteWorkflow: jest.fn(),
      duplicateWorkflow: jest.fn(() => promise),
      updateAclEntry: jest.fn(),
      deleteAclEntry: jest.fn(),
      setWorkflowPublic: jest.fn()
    }
    const onWorkflowDuplicating = jest.fn()
    const onWorkflowDuplicated = jest.fn()

    const { getByText, getByTitle } = renderWithI18n(
      <WorkflowContextMenu
        workflow={workflow}
        api={api}
        onWorkflowDuplicating={onWorkflowDuplicating}
        onWorkflowDuplicated={onWorkflowDuplicated}
        onWorkflowChanging={jest.fn()}
        onWorkflowChanged={jest.fn()}
      />
    )

    fireEvent.click(getByTitle('menu'))
    fireEvent.click(getByText('Duplicate'))

    expect(onWorkflowDuplicating).toHaveBeenCalledWith(123, {})
    expect(api.duplicateWorkflow).toHaveBeenCalledWith(123)
    expect(onWorkflowDuplicated).not.toHaveBeenCalled()
    resolvePromise({ foo: 'bar' })
    await act(async () => await null) // respond to resolvePromise()
    expect(onWorkflowDuplicated).toHaveBeenCalledWith(123, { foo: 'bar' })
  })

  it('sets workflow to public', async () => {
    const workflow = { id: 123, public: false, owner_email: 'foo@example.org', acl: [] }
    let resolvePromise
    const promise = new Promise(resolve => { resolvePromise = resolve })
    const api = {
      deleteWorkflow: jest.fn(),
      duplicateWorkflow: jest.fn(),
      updateAclEntry: jest.fn(),
      deleteAclEntry: jest.fn(),
      setWorkflowPublic: jest.fn(() => promise)
    }
    const onWorkflowChanging = jest.fn()
    const onWorkflowChanged = jest.fn()

    const { getByText, getByTitle } = renderWithI18n(
      <WorkflowContextMenu
        workflow={workflow}
        api={api}
        onWorkflowDuplicating={jest.fn()}
        onWorkflowDuplicated={jest.fn()}
        onWorkflowChanging={onWorkflowChanging}
        onWorkflowChanged={onWorkflowChanged}
      />
    )

    fireEvent.click(getByTitle('menu'))
    fireEvent.click(getByText('Share'))
    fireEvent.click(getByText('Anyone can view and duplicate this workflow, and see your email.'))

    expect(onWorkflowChanging).toHaveBeenCalledWith(123, { public: true })
    expect(api.setWorkflowPublic).toHaveBeenCalledWith(123, true)
    expect(onWorkflowChanged).not.toHaveBeenCalled()
    resolvePromise(null)
    await act(async () => await null) // respond to resolvePromise()
    expect(onWorkflowChanged).toHaveBeenCalledWith(123)
  })
})
