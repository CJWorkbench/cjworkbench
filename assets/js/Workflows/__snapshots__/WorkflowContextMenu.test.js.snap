// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkflowContextMenu renders correctly 1`] = `
<WorkflowContextMenu
  deleteWorkflow={[MockFunction]}
  duplicateWorkflow={[MockFunction]}
  workflowId={3}
>
  <UncontrolledDropdown>
    <Dropdown
      active={false}
      addonType={false}
      direction="down"
      inNavbar={false}
      isOpen={false}
      nav={false}
      setActiveFromChild={false}
      toggle={[Function]}
    >
      <Manager
        className="dropdown"
        onKeyDown={[Function]}
        tag="div"
      >
        <div
          className="dropdown"
          onKeyDown={[Function]}
        >
          <DropdownToggle
            aria-haspopup={true}
            className="context-button"
            color="secondary"
          >
            <Target
              aria-expanded={false}
              aria-haspopup={true}
              className="context-button"
              color="secondary"
              component={[Function]}
              onClick={[Function]}
            >
              <Button
                aria-expanded={false}
                aria-haspopup={true}
                className="context-button"
                color="secondary"
                innerRef={[Function]}
                onClick={[Function]}
                tag="button"
              >
                <button
                  aria-expanded={false}
                  aria-haspopup={true}
                  aria-label={null}
                  className="context-button btn btn-secondary"
                  onClick={[Function]}
                  type="button"
                >
                  <i
                    className="context-button--icon icon-more"
                  />
                </button>
              </Button>
            </Target>
          </DropdownToggle>
          <DropdownMenu
            flip={true}
            positionFixed={true}
            right={true}
            tag="div"
          >
            <div
              aria-hidden={true}
              className="dropdown-menu dropdown-menu-right"
              positionFixed={true}
              role="menu"
              tabIndex="-1"
            >
              <DropdownItem
                className="duplicate-workflow"
                onClick={[Function]}
                tag="button"
                toggle={true}
              >
                <button
                  className="duplicate-workflow dropdown-item"
                  onClick={[Function]}
                  role="menuitem"
                  tabIndex="0"
                  type="button"
                >
                  <i
                    className="icon-duplicate"
                  />
                  <span>
                    Duplicate
                  </span>
                </button>
              </DropdownItem>
              <DropdownItem
                className="delete-workflow"
                onClick={[Function]}
                tag="button"
                toggle={true}
              >
                <button
                  className="delete-workflow dropdown-item"
                  onClick={[Function]}
                  role="menuitem"
                  tabIndex="0"
                  type="button"
                >
                  <i
                    className="icon-bin"
                  />
                  <span>
                    Delete
                  </span>
                </button>
              </DropdownItem>
            </div>
          </DropdownMenu>
        </div>
      </Manager>
    </Dropdown>
  </UncontrolledDropdown>
</WorkflowContextMenu>
`;
